Modules/_sqlite/connection.c

```diff
@@ -432,48 +436,75 @@ free_callback_contexts(pysqlite_Connection *self)
 static void
 remove_callbacks(sqlite3 *db)
 {
+    if (db == NULL) {
+        return;
+    }
     sqlite3_trace_v2(db, SQLITE_TRACE_STMT, 0, 0);
     sqlite3_progress_handler(db, 0, 0, (void *)0);
     (void)sqlite3_set_authorizer(db, NULL, NULL);
 }
 
-static void
+static int
 connection_close(pysqlite_Connection *self)
 {
     if (self->db) {
         if (self->autocommit == AUTOCOMMIT_DISABLED &&
             !sqlite3_get_autocommit(self->db))
         {
-            /* If close is implicitly called as a result of interpreter
-             * tear-down, we must not call back into Python. */
-            if (_Py_IsInterpreterFinalizing(PyInterpreterState_Get())) {
-                remove_callbacks(self->db);
+            if (connection_exec_stmt(self, "ROLLBACK") < 0) {
+                return -1;
             }
-            (void)connection_exec_stmt(self, "ROLLBACK");
         }
 
-        free_callback_contexts(self);
-
         sqlite3 *db = self->db;
         self->db = NULL;
 
         Py_BEGIN_ALLOW_THREADS
         int rc = sqlite3_close_v2(db);
         assert(rc == SQLITE_OK), (void)rc;
         Py_END_ALLOW_THREADS
+
+        free_callback_contexts(self);
     }
+    return 0;
 }
 
 static void
-connection_dealloc(pysqlite_Connection *self)
+connection_finalize(PyObject *self)
 {
-    PyTypeObject *tp = Py_TYPE(self);
-    PyObject_GC_UnTrack(self);
-    tp->tp_clear((PyObject *)self);
+    pysqlite_Connection *con = (pysqlite_Connection *)self;
+    PyObject *exc = PyErr_GetRaisedException();
+
+    /* If close is implicitly called as a result of interpreter
+     * tear-down, we must not call back into Python. */
+    PyInterpreterState *interp = PyInterpreterState_Get();
+    int teardown = _Py_IsInterpreterFinalizing(interp);
+    if (teardown) {
```

Comment:
Wait, isn't it the opposite?

Reply:
Yes 😆 It was late; I'll fix it today.

Reply:
No, you're confusing me! The logic is unchanged:

If we're finalizing, we need to clear the callbacks before doing ROLLBACK (we don't want Python callbacks to trigger during interpreter teardown). Also, in case of teardown, we clear any exception instead of logging unraisable exception.